File: groundstation-frontend/src/types/obstacle-avoidance.ts
export type ObstacleSensorType = 'millimeter_wave_radar' | 'stereo_vision' | 'lidar' | 'ultrasonic'

export type AvoidanceSensitivity = 'far' | 'medium' | 'near'

export type AvoidanceStrategy = 'hover' | 'ascend_bypass' | 'retreat_bypass'

export type ObstacleDirection = 'front' | 'left' | 'right' | 'top' | 'bottom' | 'rear'

export type AvoidanceActionStatus = 'detecting' | 'triggered' | 'avoiding' | 'bypassing' | 'completed' | 'failed'

export interface ObstacleDetection {
  id: string
  uavId: string
  timestamp: number
  sensorType: ObstacleSensorType
  direction: ObstacleDirection
  distance: number
  relativeAngle: number
  obstacleSize: number
  confidence: number
  position: {
    lat: number
    lng: number
    alt: number
  }
}

export interface ObstacleAvoidanceEvent {
  id: string
  uavId: string
  timestamp: number
  detection: ObstacleDetection
  strategy: AvoidanceStrategy
  status: AvoidanceActionStatus
  startPosition: {
    lat: number
    lng: number
    alt: number
  }
  bypassPath: BypassWaypoint[]
  completedAt?: number
  failReason?: string
}

export interface BypassWaypoint {
  lat: number
  lng: number
  alt: number
  timestamp: number
  type: 'detection_point' | 'bypass_start' | 'bypass_waypoint' | 'bypass_end' | 'resume_point'
}

export interface ObstacleAvoidanceConfig {
  uavId: string
  enabled: boolean
  sensitivity: AvoidanceSensitivity
  strategy: AvoidanceStrategy
  sensorType: ObstacleSensorType
  detectionRange: number
  minObstacleSize: number
  ascendHeight: number
  retreatDistance: number
  bypassAngle: number
}

export interface ObstacleHeatmapPoint {
  lat: number
  lng: number
  alt: number
  triggerCount: number
  lastTriggerTime: number
  intensity: number
  avgDistance: number
  minDistance: number
}

export interface ObstacleAvoidanceLog {
  id: string
  uavId: string
  uavName?: string
  timestamp: number
  sensorType: ObstacleSensorType
  direction: ObstacleDirection
  distance: number
  strategy: AvoidanceStrategy
  status: AvoidanceActionStatus
  duration?: number
  position: {
    lat: number
    lng: number
    alt: number
  }
  bypassPathLength?: number
  description: string
}

export interface ObstacleAvoidanceStatistics {
  totalDetections: number
  totalAvoidanceEvents: number
  successfulAvoidances: number
  failedAvoidances: number
  avgReactionTime: number
  avgAvoidanceDuration: number
  nearestObstacleDistance: number
  strategyDistribution: Record<AvoidanceStrategy, number>
  directionDistribution: Record<ObstacleDirection, number>
  sensitivitySettings: AvoidanceSensitivity
}

export const SENSITIVITY_CONFIG: Record<AvoidanceSensitivity, { detectionRange: number; label: string; reactionDistance: number }> = {
  far: { detectionRange: 15, label: '远距 (15m)', reactionDistance: 12 },
  medium: { detectionRange: 10, label: '中距 (10m)', reactionDistance: 8 },
  near: { detectionRange: 5, label: '近距 (5m)', reactionDistance: 4 }
}

export const STRATEGY_LABELS: Record<AvoidanceStrategy, string> = {
  hover: '悬停',
  ascend_bypass: '上升绕行',
  retreat_bypass: '后退绕行'
}

export const SENSOR_TYPE_LABELS: Record<ObstacleSensorType, string> = {
  millimeter_wave_radar: '毫米波雷达',
  stereo_vision: '双目视觉',
  lidar: '激光雷达',
  ultrasonic: '超声波'
}

export const DIRECTION_LABELS: Record<ObstacleDirection, string> = {
  front: '前方',
  left: '左侧',
  right: '右侧',
  top: '上方',
  bottom: '下方',
  rear: '后方'
}

export const ACTION_STATUS_LABELS: Record<AvoidanceActionStatus, string> = {
  detecting: '检测中',
  triggered: '已触发',
  avoiding: '避障中',
  bypassing: '绕行中',
  completed: '已完成',
  failed: '失败'
}

export const ACTION_STATUS_COLORS: Record<AvoidanceActionStatus, string> = {
  detecting: '#1890ff',
  triggered: '#faad14',
  avoiding: '#fa8c16',
  bypassing: '#722ed1',
  completed: '#52c41a',
  failed: '#ff4d4f'
}

File: groundstation-backend/internal/models/obstacle_avoidance.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type ObstacleSensorType string

const (
	SensorTypeMillimeterWave ObstacleSensorType = "millimeter_wave_radar"
	SensorTypeStereoVision   ObstacleSensorType = "stereo_vision"
	SensorTypeLidar          ObstacleSensorType = "lidar"
	SensorTypeUltrasonic     ObstacleSensorType = "ultrasonic"
)

type AvoidanceSensitivity string

const (
	SensitivityFar    AvoidanceSensitivity = "far"
	SensitivityMedium AvoidanceSensitivity = "medium"
	SensitivityNear   AvoidanceSensitivity = "near"
)

type AvoidanceStrategy string

const (
	StrategyHover       AvoidanceStrategy = "hover"
	StrategyAscendBypass AvoidanceStrategy = "ascend_bypass"
	StrategyRetreatBypass AvoidanceStrategy = "retreat_bypass"
)

type ObstacleDirection string

const (
	DirectionFront  ObstacleDirection = "front"
	DirectionLeft   ObstacleDirection = "left"
	DirectionRight  ObstacleDirection = "right"
	DirectionTop    ObstacleDirection = "top"
	DirectionBottom ObstacleDirection = "bottom"
	DirectionRear   ObstacleDirection = "rear"
)

type AvoidanceActionStatus string

const (
	AvoidanceStatusDetecting AvoidanceActionStatus = "detecting"
	AvoidanceStatusTriggered AvoidanceActionStatus = "triggered"
	AvoidanceStatusAvoiding  AvoidanceActionStatus = "avoiding"
	AvoidanceStatusBypassing AvoidanceActionStatus = "bypassing"
	AvoidanceStatusCompleted AvoidanceActionStatus = "completed"
	AvoidanceStatusFailed    AvoidanceActionStatus = "failed"
)

type ObstacleAvoidanceConfig struct {
	ID             uint64              `gorm:"primaryKey;autoIncrement" json:"id"`
	UAVID          uint64              `gorm:"uniqueIndex;not null" json:"uav_id"`
	Enabled        bool                `gorm:"default:true" json:"enabled"`
	Sensitivity    AvoidanceSensitivity `gorm:"type:varchar(20);default:'medium'" json:"sensitivity"`
	Strategy       AvoidanceStrategy   `gorm:"type:varchar(20);default:'ascend_bypass'" json:"strategy"`
	SensorType     ObstacleSensorType  `gorm:"type:varchar(30);default:'millimeter_wave_radar'" json:"sensor_type"`
	DetectionRange float64             `gorm:"type:decimal(5,1);default:15" json:"detection_range"`
	MinObstacleSize float64            `gorm:"type:decimal(5,2);default:0.5" json:"min_obstacle_size"`
	AscendHeight   float64             `gorm:"type:decimal(5,1);default:5" json:"ascend_height"`
	RetreatDistance float64             `gorm:"type:decimal(5,1);default:10" json:"retreat_distance"`
	BypassAngle    float64             `gorm:"type:decimal(5,1);default:45" json:"bypass_angle"`
	CreatedAt      time.Time           `json:"created_at"`
	UpdatedAt      time.Time           `json:"updated_at"`
}

func (ObstacleAvoidanceConfig) TableName() string {
	return "obstacle_avoidance_configs"
}

type ObstacleDetectionLog struct {
	ID          uint64              `gorm:"primaryKey;autoIncrement" json:"id"`
	UAVID       uint64              `gorm:"index;not null" json:"uav_id"`
	SensorType  ObstacleSensorType  `gorm:"type:varchar(30)" json:"sensor_type"`
	Direction   ObstacleDirection   `gorm:"type:varchar(20)" json:"direction"`
	Distance    float64             `gorm:"type:decimal(8,2);not null" json:"distance"`
	RelativeAngle float64           `gorm:"type:decimal(8,2)" json:"relative_angle"`
	ObstacleSize float64            `gorm:"type:decimal(5,2)" json:"obstacle_size"`
	Confidence  float64             `gorm:"type:decimal(3,2)" json:"confidence"`
	Latitude    float64             `gorm:"type:decimal(10,7)" json:"lat"`
	Longitude   float64             `gorm:"type:decimal(10,7)" json:"lng"`
	Altitude    float64             `gorm:"type:decimal(8,2)" json:"alt"`
	CreatedAt   time.Time           `json:"created_at"`
}

func (ObstacleDetectionLog) TableName() string {
	return "obstacle_detection_logs"
}

type ObstacleAvoidanceEvent struct {
	ID             uint64               `gorm:"primaryKey;autoIncrement" json:"id"`
	UAVID          uint64               `gorm:"index;not null" json:"uav_id"`
	Strategy       AvoidanceStrategy    `gorm:"type:varchar(20);not null" json:"strategy"`
	Status         AvoidanceActionStatus `gorm:"type:varchar(20);default:'triggered'" json:"status"`
	DetectionID    uint64               `gorm:"index" json:"detection_id"`
	SensorType     ObstacleSensorType   `gorm:"type:varchar(30)" json:"sensor_type"`
	Direction      ObstacleDirection    `gorm:"type:varchar(20)" json:"direction"`
	ObstacleDistance float64            `gorm:"type:decimal(8,2)" json:"obstacle_distance"`
	StartLat       float64              `gorm:"type:decimal(10,7)" json:"start_lat"`
	StartLng       float64              `gorm:"type:decimal(10,7)" json:"start_lng"`
	StartAlt       float64              `gorm:"type:decimal(8,2)" json:"start_alt"`
	BypassPath     string               `gorm:"type:json" json:"bypass_path"`
	CompletedAt    *time.Time           `json:"completed_at"`
	FailReason     string               `gorm:"type:varchar(500)" json:"fail_reason"`
	CreatedAt      time.Time            `json:"created_at"`
	UpdatedAt      time.Time            `json:"updated_at"`
	DeletedAt      gorm.DeletedAt       `gorm:"index" json:"-"`
}

func (ObstacleAvoidanceEvent) TableName() string {
	return "obstacle_avoidance_events"
}

type ObstacleHeatmapPoint struct {
	ID              uint64    `gorm:"primaryKey;autoIncrement" json:"id"`
	UAVID           uint64    `gorm:"index;not null" json:"uav_id"`
	Latitude        float64   `gorm:"type:decimal(10,7);not null" json:"lat"`
	Longitude       float64   `gorm:"type:decimal(10,7);not null" json:"lng"`
	Altitude        float64   `gorm:"type:decimal(8,2)" json:"alt"`
	TriggerCount    int       `gorm:"default:1" json:"trigger_count"`
	LastTriggerTime time.Time `json:"last_trigger_time"`
	AvgDistance     float64   `gorm:"type:decimal(8,2)" json:"avg_distance"`
	MinDistance     float64   `gorm:"type:decimal(8,2)" json:"min_distance"`
	CreatedAt       time.Time `json:"created_at"`
	UpdatedAt       time.Time `json:"updated_at"`
}

func (ObstacleHeatmapPoint) TableName() string {
	return "obstacle_heatmap_points"
}
